state-chain/chains/src/sol/program_instructions.rs

```diff
@@ -0,0 +1,241 @@
+		base: Pubkey,
+
+		/// String of ASCII chars, no longer than `pubkey::MAX_SEED_LEN`
+		seed: String,
+
+		/// Number of bytes of memory to allocate
+		space: u64,
+
+		/// Owner program account
+		owner: Pubkey,
+	},
+
+	/// Assign account to a program based on a seed
+	///
+	/// # Account references
+	///   0. `[WRITE]` Assigned account
+	///   1. `[SIGNER]` Base account
+	AssignWithSeed {
+		/// Base public key
+		base: Pubkey,
+
+		/// String of ASCII chars, no longer than `pubkey::MAX_SEED_LEN`
+		seed: String,
+
+		/// Owner program account
+		owner: Pubkey,
+	},
+
+	/// Transfer lamports from a derived address
+	///
+	/// # Account references
+	///   0. `[WRITE]` Funding account
+	///   1. `[SIGNER]` Base for funding account
+	///   2. `[WRITE]` Recipient account
+	TransferWithSeed {
+		/// Amount to transfer
+		lamports: u64,
+
+		/// Seed to use to derive the funding account address
+		from_seed: String,
+
+		/// Owner to use to derive the funding account address
+		from_owner: Pubkey,
+	},
+
+	/// One-time idempotent upgrade of legacy nonce versions in order to bump
+	/// them out of chain blockhash domain.
+	///
+	/// # Account references
+	///   0. `[WRITE]` Nonce account
+	UpgradeNonceAccount,
+}
+
+impl SystemProgramInstruction {
+	pub fn advance_nonce_account(nonce_pubkey: &Pubkey, authorized_pubkey: &Pubkey) -> Instruction {
+		let account_metas = vec![
+			AccountMeta::new(*nonce_pubkey, false),
+			// the public key for RecentBlockhashes system variable.
+			//
+			// NOTE: According to the solana sdk, this system variable is deprecated and should not
+			// be used. However, within the sdk itself they are still using this variable in the
+			// advance_nonce_account function so we use it here aswell. This should be revisited to
+			// make sure it is ok to use it, or if there is another way to advance the nonce
+			// account.
+			AccountMeta::new_readonly(
+				Pubkey::from_str("SysvarRecentB1ockHashes11111111111111111111").unwrap(),
+				false,
+			),
+			AccountMeta::new_readonly(*authorized_pubkey, true),
+		];
+		Instruction::new_with_bincode(
+			// program id of the system program
+			Pubkey::from_str(SYSTEM_PROGRRAM_ID).unwrap(),
+			&Self::AdvanceNonceAccount,
+			account_metas,
+		)
+	}
+
+	pub fn transfer(from_pubkey: &Pubkey, to_pubkey: &Pubkey, lamports: u64) -> Instruction {
+		let account_metas =
+			vec![AccountMeta::new(*from_pubkey, true), AccountMeta::new(*to_pubkey, false)];
+		Instruction::new_with_bincode(
+			Pubkey::from_str(SYSTEM_PROGRRAM_ID).unwrap(),
+			&Self::Transfer { lamports },
+			account_metas,
+		)
+	}
+}
+
+#[derive(BorshDeserialize, BorshSerialize, Debug, Clone, PartialEq, Eq)]
+pub enum VaultProgram {
+	FetchSol { seed: Vec<u8>, bump: u8 },
+}
+
+impl VaultProgram {
+	pub fn get_instruction(self, accounts: Vec<AccountMeta>) -> Instruction {
+		let mut instruction =
+			Instruction::new_with_borsh(Pubkey::from_str(VAULT_PROGRAM).unwrap(), &self, accounts);
+		instruction.data.remove(0);
```

Comment:
What is this remove(0) for?

Reply:
This is basically because of custom instruction serialization that anchor (framework which we are using to write solana program) uses. Basically, raw solana program have indices of calls as function selectors but anchor uses the hash of the function name and so we remove the index byte and add the hash of the call as function selector.